demos/smart_classroom_demo/cpp_gapi/main.cpp

```diff
@@ -0,0 +1,395 @@
+        stream.setSource<custom::CustomCapSource>(cap);
+
+        /** Service constants **/
+        float wait_time_ms = 0.f;
+        float work_time_ms = 0.f;
+        size_t wait_num_frames = 0;
+        size_t work_num_frames = 0;
+        size_t total_num_frames = 0;
+        size_t work_time_ms_all = 0;
+        const char SPACE_KEY = 32;
+        const char ESC_KEY = 27;
+        bool monitoring_enabled = const_params.actions_type == TOP_K ? false : true;
+        cv::Size graphSize { static_cast<int>(frame_size.width / 4), 60 };
+
+        /** Presenter for rendering system parameters **/
+        Presenter presenter(FLAGS_u, frame_size.height - graphSize.height - 10, graphSize);
+
+        /** Create VideoWriter **/
+        cv::VideoWriter videoWriter;
+        if (!FLAGS_o.empty() && !videoWriter.open(FLAGS_o, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'),
+                                                  cap->fps(), frame_size)) {
+            throw std::runtime_error("Can't open video writer");
+        }
+
+        /** Result containers associated with graph output **/
+        cv::Mat out_frame, proc, top_k;
+        std::string stream_log, stat_log, det_log;
+        auto out_vector = cv::gout(out_frame);
+        if (const_params.actions_type == TOP_K) {
+            out_vector += cv::gout(top_k, proc);
+        } else {
+            out_vector += cv::gout(work_num_frames, stream_log, stat_log, det_log, proc);
+        }
+
+        /** TOP_K case starts without processing **/
+        if (const_params.actions_type != TOP_K) stream.start();
+
+        std::cout << "To close the application, press 'CTRL+C' here";
+        if (!FLAGS_no_show) {
+            std::cout << " or switch to the output window and press ESC key";
+        }
+        std::cout << std::endl;
+
+        /** Main cycle **/
+        auto started_all = std::chrono::high_resolution_clock::now();
+        while (true) {
+            auto started = std::chrono::high_resolution_clock::now();
+            char key = cv::waitKey(1);
+            presenter.handleKey(key);
+            if (key == ESC_KEY) {
+                break;
+            }
+            if (const_params.actions_type == TOP_K) {
+                if ((key == SPACE_KEY && !monitoring_enabled) ||
+                    (key == SPACE_KEY && monitoring_enabled)) {
+                    /** SPACE_KEY & monitoring_enabled trigger **/
+                    monitoring_enabled = !monitoring_enabled;
+                    const_params.draw_ptr->ClearTopWindow();
+                }
+            }
+            if (monitoring_enabled) {
+                if (!stream.running()) {
+                    /** TOP_K part. SPACE_KEY is pressed, monitoring enabled
+                     *  Compile and start graph **/
+                    cv::Mat tmp = cap->read();
+                    if (!tmp.data) break;
+                    stream.setSource<custom::CustomCapSource>(cap);
```

Comment:
I'm not sure I follow this. `CustomCapSource` already has `prep()` checking if first image isn't empty. Pluss when you do `cv::Mat tmp = cap->read();` you remove that frame from `cap`, so the image is lost.